BUG: Prevent addition overflow with TimedeltaIndex

gfyoung opened the pull request:
Expands checked-add array addition introduced in #14237 to include all other addition cases (i.e.
`TimedeltaIndex` and `Timedelta`).  Follow-up to #14453.

gfyoung committed:
BUG: Prevent addition overflow with TimedeltaIndex

Expands checked-add array addition introduced in
gh-14237 to include all other addition cases (i.e.
TimedeltaIndex and Timedelta). Follow-up to gh-14453.

In addition, move checked add function to core/algorithms.

jreback commented on pandas/core/nanops.py:
```diff
@@ -812,15 +812,21 @@ def unique1d(values):
 
 
-def _checked_add_with_arr(arr, b):
+def _checked_add_with_arr(arr, b, arr_nans=None, b_nans=None):
```
why don't we make this signature more consistent maybe
``_checked_add_with_arr(a, b, a_mask=None, b_mask=None)``

jreback replied:
btw, this function is differnt than most in ``nanops.py``. we normally actually just call ``isnull`` directly on the arrays, so we *don't* first send in the ``.view('i8')`` for datetimelikes (this is done in the function). I think you should make this more consistent e.g. ideally this signature would actually be: ``_check_add_with_arr(a, b, skipna=True)``

gfyoung replied:
Sure but you do understand that that isn't what we're getting when doing TDI addition / subtraction.  In addition, there is no need to check for `isnull` since the work already done for us during instantiation of the TDI instance.

jorisvandenbossche replied:
Making the signature `_checked_add_with_arr(a, b, a_mask=None, b_mask=None)` is OK, but I agree with @gfyoung that passing the actual masks to the function like it is now is good.

jreback replied:
the problem is future readers will just be plain confused

change this to compute the mask internally like all other functions

gfyoung replied:
Not with good documentation they won't.  I'm not convinced by your consistency argument because you're going to sacrifice performance.  Every function in `nanops.py` calls `isnull`, but we have no reason to do this because we already know the `null` values before calling the function.  The other option is that we would call the specific attribute of the TDI object internally but then that makes no other universal to all `pandas` objects.

jreback commented on pandas/core/nanops.py:
```diff
@@ -812,15 +812,21 @@ def unique1d(values):
     ----------
     arr : array addend.
     b : array or scalar addend.
+    arr_nans : array indicating which elements are NaN
```
```
a_mask : boolean array or None
    array indicated which elements to exclude from checking
```

gfyoung replied:
Done

jorisvandenbossche commented:
Can you show the results of the related benchmarks before/after PR?

gfyoung commented:
Setup script:
~~~python
import pandas as pd
import numpy as np

checked_add = pd.core.nanops._checked_add_with_arr

arr = np.arange(1000000)
arrmixed = np.array([1, -1]).repeat(500000)

arr_nan = np.random.choice([True, False], size=1000000)
arrmixed_nan = np.random.choice([True, False], size=1000000)
~~~
Timing info is with 100 loops, best of 3

On `master`:
* `checked_add(arr, 1)`: 8.39 ms / loop
* `checked_add(arr, -1)`: 8.38 ms / loop
* `checked_add(arr, 0)`: 8.37 ms / loop
* `checked_add(arr, arrpos)`: 9.53 ms / loop
* `checked_add(arr, arrneg)`: 9.34 ms / loop
* `checked_add(arr, arrmixed)`: 14.9 ms / loop

On `PR`:
* `checked_add(arr, 1)`: 8.35 ms / loop
* `checked_add(arr, -1)`: 8.34 ms / loop
* `checked_add(arr, 0)`: 8.42 ms / loop
* `checked_add(arr, arrpos)`: 10.1 ms / loop
* `checked_add(arr, arrneg)`: 9.6 ms / loop
* `checked_add(arr, arrmixed)`: 16.8 ms / loop

gfyoung commented:
@jreback , @jorisvandenbossche : How does the performance look to you both?  If it looks good, should be ready to merge if there are no other concerns.

jorisvandenbossche commented:
Those timings look good. Can you add a test for timedelta where the overflow is raised (for the cases where you added the check)

jreback commented:
you didn't address my api comments

jreback commented:
this is inconsistent with the rest of the code

gfyoung commented:
@jreback : I did address your API comments <a href="https://github.com/pandas-dev/pandas/pull/14816#discussion_r91802356">here</a>.  It also explains why I don't really think consistency is an issue here.  I don't understand why you want to make this internal function "consistent," especially when it doesn't really make sense to align it with the rest of the other functions.

gfyoung commented:
@jreback @jorisvandenbossche : Do you have anything more to add about this API consistency issue, especially in light of my response?

jorisvandenbossche commented:
As a possible way out of the discussion: @jreback, as your main point was a lack of consistency with the rest of the functions in `nanops.py`, would the current version be more OK to you when we moved it to another location (and actually, this function is not really a "nanop" anyway, it is a adapted `add` to deal with integer overflow in the first place, not NaNs). Eg `core/algorithms.py` or somewhere else.

jreback commented:
would be ok with a move to core/algorithms.py and api as currently (meaning the last iteration)

gfyoung commented:
Changed `_checked_add_with_arr` to `checked_add_with_arr` in `core/algorithms.py` since we are exposing it in `core/api.py`.  Added `whatsnew` to `v0.20.0` since this is technically an API change.

gfyoung commented:
Changed `_checked_add_with_arr` to `checked_add_with_arr` in `core/algorithms.py` since we are exposing it in `core/api.py`.  Added `whatsnew` to `v0.20.0` since this is technically an API change.

jreback commented on pandas/core/api.py:
```diff
@@ -4,7 +4,8 @@
 import numpy as np
 
-from pandas.core.algorithms import factorize, match, unique, value_counts
+from pandas.core.algorithms import (checked_add_with_arr, factorize, match,
```
NO! this is not a top-level function, nor should be actually exposed to the user

jreback replied:
the name is fine, but no reason to expose this

gfyoung replied:
> would be ok with a move to core/algorithms.py and api as currently (meaning the last iteration)

Those were your words.  Fair enough.  Will remove.

jreback commented:
@gfyoung moving to an internal pandas module and exposing as a public method are 2 very separate and distinct things.

gfyoung commented:
@jreback : Yes, I am aware, so then what did you mean by "a move to `core/algorithms` and `api`" then?

jreback commented:
api as you have currently, e.g. ``(a, b, a_mask, b_mask)`` (though I think you have something slightly different)

gfyoung commented:
@jreback : Ah, I see.  Signature is `(arr, b, arr_mask=None, b_mask=None)`.

gfyoung commented:
@jreback , @jorisvandenbossche : Moved checked add function to `core/algorithms.py`, and everything is still passing.  Ready to merge if there are no other concerns.

jorisvandenbossche commented on asv_bench/benchmarks/algorithms.py:
```diff
@@ -17,13 +17,16 @@ def setup(self):
         self.arrmixed = np.array([1, -1]).repeat(500000)
 
+        self.arr_nan = np.random.choice([True, False], size=1000000)
+        self.arrmixed_nan = np.random.choice([True, False], size=1000000)
```
What's the difference between those two?

gfyoung replied:
Those are the masks for `self.arr` and `self.arrmixed` respectively.

jorisvandenbossche replied:
But they are defined exactly the same, that's the reason I am confused on why two different variables are needed and the mask is not just reused. (but they are of course both random, so not exactly the same)

gfyoung replied:
I fail to see your confusion here.  Why re-use the mask when that won't be the case in reality (i.e. the masks will likely be different in most cases)?

jorisvandenbossche replied:
yeah, no problem (I also merged BTW :-)), I was just skimming the code, and saw two apparantly identical lines of code, hence my comment (which was not needed)

jorisvandenbossche commented on asv_bench/benchmarks/algorithms.py:
```diff
@@ -64,6 +67,16 @@ def time_add_overflow_neg_arr(self):
+
+    def time_add_overflow_both_arg_nan(self):
+        self.checked_add(self.arr, self.arrmixed, arr_mask=self.arr_nan,
+                         b_mask=self.arrmixed_arr_nan)
```
`arrmixed_arr_nan` should this be `arrmixed_nan` ? (above as well)

gfyoung replied:
Good catch.  Will fix.

jorisvandenbossche commented on doc/source/whatsnew/v0.19.2.txt:
```diff
@@ -56,6 +56,7 @@ Bug Fixes
 
 
 
+Bug in ``TimedeltaIndex`` addition where overflow was being allowed without error (:issue:`14816`)
```
Can you move to 0.20.0 ?

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -40,6 +41,94 @@
+        Helper function to broadcast arrays / scalars to the desired shape.
+
+        This function is compatible with different versions of NumPy and is
+        implemented for performance reasons.
```
"implemented for performance reasons" is a bit unclear I find. We broadcast for performance reasons, but the function is not implemented for that reason, that is to broadcast for older numpy versions. 
I would just keep your original comment about performance and broadcasting

gfyoung replied:
Fair enough.  Done.

jorisvandenbossche commented on pandas/core/algorithms.py:
```diff
@@ -40,6 +41,94 @@
 # top-level algos #
 # --------------- #
 
+def checked_add_with_arr(arr, b, arr_mask=None, b_mask=None):
```
Can you put this function somewhere more below in this file? As it does not belong to the "top-level algos" (the title just above this)

jorisvandenbossche reviewed (commented):
* Can you add a test for timedelta where the overflow is raised? (for the cases where you added the check)

* Can you show the output of the `Algorithms` benchmark, just to be sure they run?

gfyoung commented:
@jorisvandenbossche :

~~~
· Discovering benchmarks
· Running 16 total benchmarks (1 commits * 1 environments * 16 benchmarks)
[  0.00%] ·· Building for existing-py_home_user_miniconda3_envs_pandas-dev_bin_python
[  0.00%] ·· Benchmarking existing-py_home_user_miniconda3_envs_pandas-dev_bin_python
[  6.25%] ··· Running ...ms.Algorithms.time_add_overflow_both_arg_nan    16.24ms
[ 12.50%] ··· Running ...s.Algorithms.time_add_overflow_first_arg_nan    15.98ms
[ 18.75%] ··· Running ...ithms.Algorithms.time_add_overflow_mixed_arr    16.67ms
[ 25.00%] ··· Running algorithms.Algorithms.time_add_overflow_neg_arr     8.93ms
[ 31.25%] ··· Running ...thms.Algorithms.time_add_overflow_neg_scalar     8.02ms
[ 37.50%] ··· Running algorithms.Algorithms.time_add_overflow_pos_arr     9.03ms
[ 43.75%] ··· Running ...thms.Algorithms.time_add_overflow_pos_scalar     8.07ms
[ 50.00%] ··· Running ....Algorithms.time_add_overflow_second_arg_nan    15.96ms
[ 56.25%] ··· Running ...hms.Algorithms.time_add_overflow_zero_scalar     8.12ms
[ 62.50%] ··· Running algorithms.Algorithms.time_duplicated_float        18.64ms
[ 68.75%] ··· Running algorithms.Algorithms.time_duplicated_int          15.32ms
[ 75.00%] ··· Running ...rithms.Algorithms.time_duplicated_int_unique    50.73μs
[ 81.25%] ··· Running algorithms.Algorithms.time_factorize_float          7.59ms
[ 87.50%] ··· Running algorithms.Algorithms.time_factorize_int            7.49ms
[ 93.75%] ··· Running algorithms.Algorithms.time_factorize_string        29.15ms
[100.00%] ··· Running algorithms.Algorithms.time_match_strings            1.91ms
~~~

gfyoung commented:
@jreback , @jorisvandenbossche : Addressed all of the comments and ran the benchmarks as requested.  Everything is green, so ready to merge if there is nothing else.

jorisvandenbossche commented:
@gfyoung thanks!
